cmd/gitserver/server/common/queue.go

```diff
@@ -3,45 +3,154 @@ package common
+	}, metricLabels)
+
+	queueProcessingTime = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "src_gitserver_generic_queue_processing_time_seconds",
+	}, metricLabels)
 )
 
+var registerMetricsOnce = new(sync.Once)
+
+func registerMetrics(observationCtx *observation.Context) {
+	registerMetricsOnce.Do(func() {
+		observationCtx.Registerer.MustRegister(
+			queueLength,
+			queueEnqueuedTotal,
+			queueWaitTime,
+			queueProcessingTime,
+		)
+	})
+}
+
+type Jobber interface {
+	// Identifier returns a string that will help identify the job. It is **NOT** a unique ID for
+	// every job, rather it must return the same string if the same job is pushed to the queue
+	// twice.
+	//
+	// This is required in tracking metrics like wait time and processing time of a job.
+	Identifier() string
+
+	// GetPushedAt returns the pushedAt time from the JobMetadata.
+	//
+	// If it is invoked before SetPushedAt (see below), it will return the zero value of time.Time.
+	GetPushedAt() time.Time
+
+	// SetPushedAt sets the pushedAt time on the JobMetadata. It is idempotent.
+	//
+	// Invoking this the first time will set the current time and subsequent calls to this method
+	// will be a NOOP.
+	SetPushedAt()
+}
+
+// JobMetadata provides an API for managing metadata of a job. This is useful in metrics collection.
+//
+// Implementations of Jobber should embed this struct.
+type JobMetadata struct {
+	pushedAt     time.Time
+	pushedAtOnce sync.Once
+}
+
+func (jm *JobMetadata) GetPushedAt() time.Time {
+	return jm.pushedAt
+}
+func (jm *JobMetadata) SetPushedAt() {
+	jm.pushedAtOnce.Do(func() {
+		jm.pushedAt = time.Now()
+	})
+}
+
 // Queue is a threadsafe FIFO queue.
-type Queue[T any] struct {
-	mu   sync.Mutex
+type Queue[T Jobber] struct {
+	*metrics
+
 	jobs *list.List
 
+	mu sync.Mutex
+
+	// FIXME: Make these private.
+	// Coming soon in a follow up PR.
 	Mutex sync.Mutex
 	Cond  *sync.Cond
 }
 
 // NewQueue initializes a new Queue.
-func NewQueue[T any](jobs *list.List) *Queue[T] {
+func NewQueue[T Jobber](obctx *observation.Context, name string, jobs *list.List) *Queue[T] {
 	q := Queue[T]{jobs: jobs}
 	q.Cond = sync.NewCond(&q.Mutex)
 
+	// Register the metrics the first time this queue is used.
+	registerMetrics(obctx)
+
+	// Setup the metrics for this specific instance of the queue.
+	q.metrics = &metrics{
+		length:         queueLength.WithLabelValues(name),
+		enqueuedTotal:  queueEnqueuedTotal.WithLabelValues(name),
+		waitTime:       queueWaitTime.WithLabelValues(name),
+		processingTime: queueProcessingTime.WithLabelValues(name),
+	}
+
 	return &q
 }
 
 // Push will queue the job to the end of the queue.
-func (q *Queue[T]) Push(job *T) {
+func (q *Queue[T]) Push(job T) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
 	q.jobs.PushBack(job)
 	q.Cond.Signal()
+
+	// Set the push time on the job's metadata. This will be used to observe the total wait time in
+	// queue when this job is eventually popped.
+	job.SetPushedAt()
```

Comment:
FYI if you want to remove the interface Jobber entirely you can remove SetPushedAt by having

```go
type queueItem[T] struct {
    value T
    pushedTime time.Time
}
```

then

```suggestion
    q.jobs.PushBack(queueItem[T]{
      value: job,
      pushedTime: time.Now(),
    })
	q.Cond.Signal()
```

Reply:
Makes sense. I've dropped the interface and it looks much simpler now. Thank you. :)